App/In-App Actions/URLMenuPresenter.swift

```diff
@@ -441,14 +442,75 @@ final class URLMenuPresenter: NSObject {
             return true
         }
         
-        for case .spoiledVideo(frame: let frame, url: let unresolved) in elements {
+        for case .spoiledVideo(frame: _, url: let unresolved) in elements {
             if let resolved = URL(string: unresolved.absoluteString, relativeTo: ForumsClient.shared.baseURL) {
-                let presenter = URLMenuPresenter(videoURL: resolved)
-                presenter.present(fromViewController: presentingViewController, fromRect: frame, inView: renderView)
+                let actionSheet = UIAlertController.makeActionSheet()
+                
+                actionSheet.addAction(.init(title: "Copy URL", style: .default, handler: { _ in
+                    UIPasteboard.general.coercedURL = resolved
+                }))
+      
+                let path = resolved.path.lowercased()
+                if (path.hasSuffix(".mp4") || path.hasSuffix(".webm") || path.hasSuffix(".gifv")) {
+                
+                actionSheet.addAction(.init(title: "Save video", style: .default, handler: { _ in
+                    let randomURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!.appendingPathComponent("\(resolved.lastPathComponent)")
+                  
+                    PHPhotoLibrary.requestAuthorization({ _ in })
+
+                    let alertError: (ErrorWorkaround) -> Void = {error in
+                        let title = "Error"
+                        let message = "Failed to save"
+                        DispatchQueue.main.async {
+                            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+                            alert.addAction(UIAlertAction(title: "OK", style: UIAlertAction.Style.cancel, handler: nil))
+                            presentingViewController.present(alert, animated: true)
+                        }
+                    }
+                    
+                    URLSession.shared.dataTask(with: resolved) { (data: Data?, response: URLResponse?, error: Error?) in
```

Comment:
Latest commit now uses downloadTask, though I could not find any way to get around saving the file before adding it to Photos.
It's better now regardless and I added a progress spinner while downloading, since there was no indication that something was happening before

Reply:
We'll end up with a file somewhere and I think that's actually better than loading it into memory. Not that I expect these video files to be absolutely massive, but it'd still be a waste.